File: C++/2-Medium/00241._为运算表达式设计优先级/00241._为运算表达式设计优先级.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// 函数声明：为了递归调用
vector<int> diffWaysToCompute(string input);

// 主要的递归函数
vector<int> compute(string& input, int start, int end) {
    vector<int> ways;
    // 遍历字符串，寻找运算符
    for (int i = start; i < end; ++i) {
        char c = input[i];
        if (c == '+' || c == '-' || c == '*') {
            // 对于每个运算符，计算左右两边的所有可能结果
            vector<int> left = compute(input, start, i);
            vector<int> right = compute(input, i + 1, end);
            // 组合左右两边的结果
            for (int l : left) {
                for (int r : right) {
                    switch (c) {
                        case '+': ways.push_back(l + r); break;
                        case '-': ways.push_back(l - r); break;
                        case '*': ways.push_back(l * r); break;
                    }
                }
            }
        }
    }
    // 如果没有运算符，直接返回数字
    if (ways.empty()) ways.push_back(stoi(input.substr(start, end - start)));
    return ways;
}

// 工具函数，处理输入字符串并调用递归函数
vector<int> diffWaysToCompute(string input) {
    return compute(input, 0, input.size());
}

int main() {
    // 示例测试
    string input = "2*3-4*5";
    vector<int> result = diffWaysToCompute(input);
    cout << "Possible results: ";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: Python/2-Medium/00241._为运算表达式设计优先级/00241._为运算表达式设计优先级.py
from typing import List

def diffWaysToCompute(input: str) -> List[int]:
    """
    为运算表达式设计优先级，找出所有可能的结果。
    
    参数:
        input (str): 输入的运算表达式。
        
    返回:
        List[int]: 所有可能的运算结果。
        
    方法:
        使用分治策略，递归处理表达式的每个运算符。
    """
    
    # 判断是否为纯数字，如果是，直接返回该数字作为结果的列表
    if input.isdigit():
        return [int(input)]
    
    results = []
    # 遍历表达式的每个字符
    for i, char in enumerate(input):
        # 当前字符如果是运算符
        if char in "+-*":
            # 分治：分别计算左右两部分的所有可能结果
            left = diffWaysToCompute(input[:i])
            right = diffWaysToCompute(input[i+1:])
            
            # 合并结果：根据当前运算符合并左右两部分的结果
            for l in left:
                for r in right:
                    if char == '+':
                        results.append(l + r)
                    elif char == '-':
                        results.append(l - r)
                    else: # char == '*'
                        results.append(l * r)
                        
    return results

# 测试代码
expressions = ["2-1-1", "2*3-4*5"]
for expr in expressions:
    print(f"表达式 '{expr}' 的所有可能结果: {diffWaysToCompute(expr)}")